codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -806,7 +808,9 @@ void doBeginRead() {
                     }
                     // We need to double check that there is nothing left to flush such as a
                     // window update frame.
-                    flush();
+                    if (windowUpdateFrameWritten) {
+                        flush();
```

Comment:
if stream state is `CLOSE`, it should not need to be `flush`, right? 
````java
if (!Http2Stream.State.CLOSED.equals(stream.state())) {
    flush();
}
````

Reply:
well I think it might still need to do to ensure connection window updates are flushed ?

Reply:
@normanmaurer for ST this change won't make any difference because we manage flushes externally. But I can imagine that there is a risk to break users who rely on the current behavior (flush before closing the stream). Maybe it's worth adding an opt-in property to control this change.

Reply:
I think this checking of stream state is sufficient. The flow controller does at least profess to free any remaining bytes on stream closure and thus ignores any more stream flow window updates [here](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java#L185-L187).